File: pubFinderFrontend/src/app/client/modules/service/publication.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { UserAuthService } from '../../../login/service/user-auth.service';
import { Publication } from '../../../Models/Publication';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PublicationService {

  constructor(private httpClient:HttpClient,private userAuth:UserAuthService) { }
  url="http://localhost:8085/api/v1/publication/";


  addPublication(publicationAddDTO:Publication): Observable<Publication> {
    const token = this.userAuth.getToken();
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
    return this.httpClient.post<any>(this.url+'add',publicationAddDTO,{headers});

  }

  updatePubToAddImage(idPost:number,image:File): Observable<Publication> {
    const formData = new FormData();
    formData.append('image', image);
    return this.httpClient.put<any>(this.url + 'updatePubToAddImage?pubId='+`${idPost}`,formData);
  }


  getPublications(){
    const token = this.userAuth.getToken();
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
    return this.httpClient.get<Publication[]>(this.url+"getPublications",{headers});

  }
  getPublicationForAdmin(pageNo:number,size:number){
    const token = this.userAuth.getToken();
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
    let url = `${this.url}getPublicationForAdmin?`;
    url += `pageNo=${pageNo}&size=${size}`;
    return this.httpClient.get<any>(url,{headers});
  }

  updateStatusOfPostByAdmin(publication:Publication,idPost:number){
    const token = this.userAuth.getToken();
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
    return this.httpClient.put<Publication>(this.url+"updateStatusOfPostByAdmin?idPost="+`${idPost}`,publication,{headers});
  }


 deletePost(idPost:Number){
  const token = this.userAuth.getToken();
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
    return this.httpClient.delete<Publication>(this.url+"deletePost?idPost="+`${idPost}`,{headers});
 }

 getMyPublications(){
  const token = this.userAuth.getToken();
  const headers = new HttpHeaders({
    'Authorization': `Bearer ${token}`
  });
  return this.httpClient.get<Publication[]>(this.url+"getMyPublications",{headers});
 }
 ListsOfProductByCategoryHomePage(){
  return this.httpClient.get<Map<string, any>>(this.url+"ListsOfProductByCategoryHomePage");
 }
 retreivePubByCountryAndCityAndCategory(country: String, city: String, category: String, pageNo: number, size: number){
  let url = `${this.url}retreivePubByCountryAndCityAndCategory?`;
  // Append country parameter if provided
  if (country != null) {
    url += `country=${country}&`;
  }

  // Append city parameter if provided
  if (city != null) {
    url += `city=${city}&`;
  }

  // Append category parameter if provided
  if (category != null) {
    url += `category=${category}&`;
  }

  // Append pageNo and size parameters
  url += `pageNo=${pageNo}&size=${size}`;
  // Log the URL
  // Make HTTP GET request with constructed URL
  return this.httpClient.get<any>(url);
 }
 getPublicationWithId(idPub:number){
  return this.httpClient.get<Publication>(this.url+'getPublicationWithId?idPublication='+`${idPub}`)
 }
 country!:any;
 region!:any;
 category!:any;


}

File: PubFinder/src/main/java/com/example/pubfinder/service/impl/PublicationService.java
package com.example.pubfinder.service.impl;

import com.example.pubfinder.dao.mapper.PublicationMapper;
import com.example.pubfinder.dto.PublicationAddDTO;
import com.example.pubfinder.dto.PublicationDTO;
import com.example.pubfinder.dto.ResponseDto;
import com.example.pubfinder.dto.ResponseModel;
import com.example.pubfinder.model.Publication;
import com.example.pubfinder.model.User;
import com.example.pubfinder.repository.PublicationRepositroy;
import com.example.pubfinder.service.PublicationIService;
import com.example.pubfinder.specifications.PublicationSpecification;
import com.example.pubfinder.utility.ResponseUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class PublicationService implements PublicationIService {
    @Autowired
    PublicationRepositroy publicationRepositroy;
    @Autowired
    PublicationMapper publicationMapper;
    @Autowired
    SessionService sessionService;
    @Autowired
    ResponseUtil responseUtil;
    @Override
    public PublicationDTO addPublication(PublicationAddDTO publicationAddDTO) {
        User user=sessionService.getUserBySession().get();
        System.out.println(user.getEmail());
        Publication publication=publicationMapper.mapToPublication(publicationAddDTO);
        publication.setCreatedAt(LocalDateTime.now());
        publication.setEnabled(false);
        publication.setUser(user);
        publicationRepositroy.save(publication);
        return publicationMapper.mapToPublicationDTO(publication);
    }

    @Override
    public PublicationDTO updatePubToAddImage(@RequestParam long pubId, @RequestParam MultipartFile image) throws IOException {
        Publication publication=publicationRepositroy.findById(pubId).get();
        publication.setImage(image.getBytes());
        publicationRepositroy.save(publication);
        return publicationMapper.mapToPublicationDTO(publication);
    }

    @Override
    public List<PublicationDTO> getPublications() {
        // Retrieve publications from the repository
        List<Publication> publications = (List<Publication>) publicationRepositroy.findAll();

        // Sort publications by created date in descending order
        Collections.sort(publications, Comparator.comparing(Publication::getCreatedAt).reversed());

        // Map sorted publications to DTOs
        List<PublicationDTO> publicationDTOS = new ArrayList<>();
        for (Publication p : publications) {
            publicationDTOS.add(publicationMapper.mapToPublicationDTO(p));
        }
        return publicationDTOS;
    }

    @Override
    public  ResponseModel<PublicationDTO> getPublicationForAdmin(Pageable pageable){
        Page<Publication> publications=publicationRepositroy.findAll(pageable);
        return buildResponse(publications);
    }

    @Override
    public PublicationDTO updateStatusOfPostByAdmin(PublicationAddDTO publicationAddDTO,long idPost) {
        Publication publication=publicationRepositroy.findById(idPost).get();
        if(publicationAddDTO.isEnabled()) {
            publication.setEnabled(true);
            publicationRepositroy.save(publication);
        }
        else{
            publication.setEnabled(false);
            publicationRepositroy.save(publication);
        }
        return publicationMapper.mapToPublicationDTO(publication);
    }

    @Override
    public ResponseDto deletePost(long idPost) {
        publicationRepositroy.deleteById(idPost);
        return responseUtil.createResponse("post deleted", "SUCCESS", "Your post deleted with succes.");
    }

    @Override
    public List<PublicationDTO> getMyPublications() {
        User user=sessionService.getUserBySession().get();
        List<Publication> publications=publicationRepositroy.findMyPubl(user.getIdUser());
        List<PublicationDTO> publicationDTOS=new ArrayList<>();
        for (Publication p:publications) {
            publicationDTOS.add(publicationMapper.mapToPublicationDTO(p));
        }
        return publicationDTOS;
    }

    @Override
    public PublicationDTO getPublicationWithId(long idPublication) {
        Publication publication=publicationRepositroy.findById(idPublication).get();
        return publicationMapper.mapToPublicationDTO(publication);
    }
    private PublicationDTO convertToDto(Publication publication) {
          PublicationDTO publicationDTO=publicationMapper.mapToPublicationDTO(publication);
          return  publicationDTO;
    }
    @Override
    public ResponseEntity<Map<String, List<PublicationDTO>>> ListsOfProductByCategoryHomePage() {
        Map<String, List<PublicationDTO>> listProduct = new HashMap<>();
        Page<Publication> pageByCategoryProduct = publicationRepositroy.findPubByCategoryProduct(PageRequest.of(0, 4));
        Page<Publication> pageByCategoryClothes = publicationRepositroy.findPubByCategoryClothes(PageRequest.of(0, 4));
        Page<Publication> pageByCategoryFoods = publicationRepositroy.findPubByCategoryFoods(PageRequest.of(0, 4));
        Page<Publication> pageByCategoryServices = publicationRepositroy.findPubByCategoryServices(PageRequest.of(0, 4));
        Page<Publication> pageByAll = publicationRepositroy.findPubAll(PageRequest.of(0, 8));


        // Convert the retrieved publications to PublicationDTO objects
        List<PublicationDTO> allproductDTOs = pageByAll.getContent().stream()
                .map(this::convertToDto) // Assuming you have a method to convert Publication to PublicationDTO
                .collect(Collectors.toList());
        // Convert the retrieved publications to PublicationDTO objects
        List<PublicationDTO> productDTOs = pageByCategoryProduct.getContent().stream()
                .map(this::convertToDto) // Assuming you have a method to convert Publication to PublicationDTO
                .collect(Collectors.toList());
        // Convert the retrieved publications to PublicationDTO objects
        List<PublicationDTO> productDTOsByClothes = pageByCategoryClothes.getContent().stream()
                .map(this::convertToDto) // Assuming you have a method to convert Publication to PublicationDTO
                .collect(Collectors.toList());
        // Convert the retrieved publications to PublicationDTO objects
        List<PublicationDTO> productDTOsByFoods = pageByCategoryFoods.getContent().stream()
                .map(this::convertToDto) // Assuming you have a method to convert Publication to PublicationDTO
                .collect(Collectors.toList());
        // Convert the retrieved publications to PublicationDTO objects
        List<PublicationDTO> productDTOsByServices = pageByCategoryServices.getContent().stream()
                .map(this::convertToDto) // Assuming you have a method to convert Publication to PublicationDTO
                .collect(Collectors.toList());
        listProduct.put("Product", productDTOs);
        listProduct.put("Clothes", productDTOsByClothes);
        listProduct.put("Foods",  productDTOsByFoods);
        listProduct.put("Services",  productDTOsByServices);
        listProduct.put("all8Products",  allproductDTOs);
        return ResponseEntity.ok(listProduct);
    }

    @Override
    public ResponseModel<PublicationDTO> retreivePubByCountryAndCityAndCategory(String country, String city, String category, Pageable pageable) {
       final Specification<Publication> specification = PublicationSpecification.searchPublicationByManyConditions(country,city,category,true);
       Page<Publication> publcationList;
        publcationList = this.publicationRepositroy.findAll(specification, pageable);
        return buildResponse(publcationList);
    }
    private ResponseModel<PublicationDTO> buildResponse(Page<Publication> pubPage) {
        List<PublicationDTO> listOfPub = pubPage.toList()
                .stream()
                .map(publicationMapper::mapToPublicationDTO)
                .collect(Collectors.toList());

        return ResponseModel.<PublicationDTO>builder()
                .pageNo(pubPage.getNumber())
                .pageSize(pubPage.getSize())
                .totalElements(pubPage.getTotalElements())
                .totalPages(pubPage.getTotalPages())
                .data(listOfPub)
                .isLastPage(pubPage.isLast())
                .build();
    }
}
